sys/dev/drm/core/drm_atomic_uapi.c

```diff
@@ -0,0 +1,1471 @@
+		fence = drm_writeback_get_out_fence(wb_conn);
+		if (!fence)
+			return -ENOMEM;
+
+		ret = setup_out_fence(&f[(*num_fences)++], fence);
+		if (ret) {
+			dma_fence_put(fence);
+			return ret;
+		}
+
+		conn_state->writeback_job->out_fence = fence;
+	}
+
+	/*
+	 * Having this flag means user mode pends on event which will never
+	 * reach due to lack of at least one CRTC for signaling
+	 */
+	if (c == 0 && (arg->flags & DRM_MODE_PAGE_FLIP_EVENT))
+		return -EINVAL;
+
+	return 0;
+}
+
+static void complete_signaling(struct drm_device *dev,
+			       struct drm_atomic_state *state,
+			       struct drm_out_fence_state *fence_state,
+			       unsigned int num_fences,
+			       bool install_fds)
+{
+	struct drm_crtc *crtc;
+	struct drm_crtc_state *crtc_state;
+	int i;
+
+	if (install_fds) {
+		for (i = 0; i < num_fences; i++)
+			fd_install(fence_state[i].fd,
+				   fence_state[i].sync_file->sf_file);
+
+		kfree(fence_state);
+		return;
+	}
+
+	for_each_new_crtc_in_state(state, crtc, crtc_state, i) {
+		struct drm_pending_vblank_event *event = crtc_state->event;
+		/*
+		 * Free the allocated event. drm_atomic_helper_setup_commit
+		 * can allocate an event too, so only free it if it's ours
+		 * to prevent a double free in drm_atomic_state_clear.
+		 */
+		if (event && (event->base.fence || event->base.file_priv)) {
+			drm_event_cancel_free(dev, &event->base);
+			crtc_state->event = NULL;
+		}
+	}
+
+	if (!fence_state)
+		return;
+
+	for (i = 0; i < num_fences; i++) {
+		if (fence_state[i].sync_file)
+			fput(fence_state[i].sync_file->sf_file);
+		if (fence_state[i].fd >= 0)
+			put_unused_fd(fence_state[i].fd);
+
+		/* If this fails log error to the user */
+		if (fence_state[i].out_fence_ptr &&
+		    put_user(-1, fence_state[i].out_fence_ptr))
+			DRM_DEBUG_ATOMIC("Couldn't clear out_fence_ptr\n");
+	}
+
+	kfree(fence_state);
+}
+
+int drm_mode_atomic_ioctl(struct drm_device *dev,
+			  void *data, struct drm_file *file_priv)
+{
+	struct drm_mode_atomic *arg = data;
+#ifdef COMPAT_FREEBSD64
+	struct drm_mode_atomic64 *arg64;
+	struct drm_mode_atomic local_arg;
+#endif
+	uint32_t __user * __capability objs_ptr;
+	uint32_t __user * __capability count_props_ptr;
+	uint32_t __user * __capability props_ptr;
+	uint64_t __user * __capability prop_values_ptr;
+	unsigned int copied_objs, copied_props;
+	struct drm_atomic_state *state;
+	struct drm_modeset_acquire_ctx ctx;
+	struct drm_out_fence_state *fence_state;
+	int ret = 0;
+	unsigned int i, j, num_fences;
+
+	/* disallow for drivers not supporting atomic: */
+	if (!drm_core_check_feature(dev, DRIVER_ATOMIC))
+		return -EOPNOTSUPP;
+
+#ifdef COMPAT_FREEBSD64
+	if (!SV_CURPROC_FLAG(SV_CHERI)) {
+		arg64 = (struct drm_mode_atomic64 *)data;
+		arg = &local_arg;
+		CP(*arg64, *arg, flags);
+		CP(*arg64, *arg, count_objs);
+		CP(*arg64, *arg, user_data);
+		CP(*arg64, *arg, reserved);
+		arg->objs_ptr = (uintcap_t)__USER_CAP(
+		    arg64->objs_ptr,
+		    arg64->count_objs * sizeof(uint32_t));
```

Comment:
Here and similarly below, `sizeof(*objs_ptr)` might be better since the type of `objs_ptr` must be the same as the type referenced by `arg->objs_ptr`.

Reply:
arg64->objs_ptr is a 64 bit integer comes from userspace that is a "pointer" to an array of 32 bit objects.
args->objs_ptr is a kernel capability kuint64cap_t.